File: S4-Systems-for-Design-and-Impl/Labs/Lab4/ProiectCSharp/ProiectCSharp/repository/database/TripDBRepository.cs
using log4net;
using System.Collections.Generic;
using System.Data;
using log4net.Repository.Hierarchy;
using ProiectCSharp.domain;

namespace ProiectCSharp.repository.database;

public class TripDBRepository: ITripRepository
{
    private static readonly ILog logger = LogManager.GetLogger(typeof(TripDBRepository));

	public TripDBRepository() {
        logger.Info("Initializing TripDBRepository");
    }
    
    public Trip FindOne(long id)
    {
        logger.Info("Finding Trip with id " + id);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "SELECT * FROM Trips WHERE id = @id";
            IDataParameter idParam = command.CreateParameter();
            idParam.ParameterName = "@id";
            idParam.Value = id;
            command.Parameters.Add(idParam);
            using (IDataReader result = command.ExecuteReader())
            {
                if (result.Read())
                {
                    string destination = result.GetString(1);
                    DateTime departureTime = result.GetDateTime(2);
                    long numberOfSeats = result.GetInt64(3);
                    Trip trip = new Trip(destination, departureTime, numberOfSeats);
                    trip.Id = id;
                    return trip;
                }
            }
        }
        return null;
    }
    
    public IEnumerable<Trip> FindAll()
    {
        logger.Info("Finding all Trips");
        IDbConnection connection = DbUtils.getConnection();
        List<Trip> trips = new List<Trip>();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "SELECT * FROM Trips";
            using (IDataReader result = command.ExecuteReader())
            {
                while (result.Read())
                {
                    long id = result.GetInt64(0);
                    string destination = result.GetString(1);
                    DateTime departureTime = result.GetDateTime(2);
                    long numberOfSeats = result.GetInt64(3);
                    Trip trip = new Trip(destination, departureTime, numberOfSeats);
                    trip.Id = id;
                    trips.Add(trip);
                }
            }
        }
        return trips;
    }
    
    public Trip Save(Trip trip)
    {
        logger.Info("Saving Trip " + trip);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "INSERT INTO Trips (destination, dateDeparture, noOfAvailableSeats) VALUES (@destination, @dateDeparture, @noOfAvailableSeats)";
            IDataParameter destinationParam = command.CreateParameter();
            destinationParam.ParameterName = "@destination";
            destinationParam.Value = trip.Destination;
            command.Parameters.Add(destinationParam);
            IDataParameter departureParam = command.CreateParameter();
            departureParam.ParameterName = "@dateDeparture";
            departureParam.Value = trip.DateDeparture;
            command.Parameters.Add(departureParam);
            IDataParameter seatsParam = command.CreateParameter();
            seatsParam.ParameterName = "@noOfAvailableSeats";
            seatsParam.Value = trip.NoOfAvailableSeats;
            command.Parameters.Add(seatsParam);
            command.ExecuteNonQuery();
        }
        return trip;
    }
    
    public Trip Delete(long id)
    {
        logger.Info("Deleting Trip with id " + id);
        Trip trip = FindOne(id);
        if (trip != null)
        {
            IDbConnection connection = DbUtils.getConnection();
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "DELETE FROM Trips WHERE id = @id";
                IDataParameter idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);
                command.ExecuteNonQuery();
            }
        }
        return trip;
    }
    
    public Trip Update(Trip trip)
    {
        logger.Info("Updating Trip " + trip);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "UPDATE Trips SET destination = @destination, dateDeparture = @dateDeparture, noOfAvailableSeats = @noOfAvailableSeats WHERE id = @id";
            IDataParameter destinationParam = command.CreateParameter();
            destinationParam.ParameterName = "@destination";
            destinationParam.Value = trip.Destination;
            command.Parameters.Add(destinationParam);
            IDataParameter departureParam = command.CreateParameter();
            departureParam.ParameterName = "@dateDeparture";
            departureParam.Value = trip.DateDeparture;
            command.Parameters.Add(departureParam);
            IDataParameter seatsParam = command.CreateParameter();
            seatsParam.ParameterName = "@noOfAvailableSeats";
            seatsParam.Value = trip.NoOfAvailableSeats;
            command.Parameters.Add(seatsParam);
            IDataParameter idParam = command.CreateParameter();
            idParam.ParameterName = "@id";
            idParam.Value = trip.Id;
            command.Parameters.Add(idParam);
            command.ExecuteNonQuery();
        }
        return trip;
    }
}

File: S4-Systems-for-Design-and-Impl/Labs/Lab11/ProiectJava/TransportPersistence/src/main/java/transport/persistence/repository/TripDBRepository.java
package transport.persistence.repository;

import org.hibernate.Session;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Repository;
import transport.model.Trip;
import transport.persistence.ITripRepository;

import java.io.IOException;
import java.sql.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

@Repository
public class TripDBRepository implements ITripRepository {

    private JdbcUtils dbUtils;
    //private static final Logger logger = LogManager.getLogger();

    public TripDBRepository(Properties props) {
        //logger.info("Initializing TripDBRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
    }

    @Autowired
    public TripDBRepository(
                @Value("${jdbc.driver}") String driver,
                @Value("${jdbc.url}") String url,
                @Value("${jdbc.user}") String user,
                @Value("${jdbc.pass}") String pass) {
            Properties props = new Properties();
            props.setProperty("transport.jdbc.driver", driver);
            props.setProperty("transport.jdbc.url", url);
            props.setProperty("transport.jdbc.user", user);
            props.setProperty("transport.jdbc.pass", pass);
//            logger.info("Initializing repository.database.ChallengeDBRepository with properties: {} ", props);
            dbUtils = new JdbcUtils(props);
    }

    @Override
    public Optional<Trip> findOne(Long id) {
        // method to find a trip by id
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Trips where id=?")) {
            preStmt.setLong(1, id);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    String destination = result.getString("destination");
                    Timestamp timestamp = result.getTimestamp("dateDeparture");
                    LocalDateTime departureTime = timestamp.toLocalDateTime();
                    Long numberOfSeats = result.getLong("noOfAvailableSeats");
                    Trip trip = new Trip(destination, departureTime, numberOfSeats);
                    trip.setId(id);
                    return Optional.of(trip);
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Iterable<Trip> findAll() {
        // method to find all trips
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Trip> trips = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Trips")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Long id = result.getLong("id");
                    String destination = result.getString("destination");
                    Timestamp timestamp = result.getTimestamp("dateDeparture");
                    LocalDateTime departureTime = timestamp.toLocalDateTime();
                    Long numberOfSeats = result.getLong("noOfAvailableSeats");
                    Trip trip = new Trip(destination, departureTime, numberOfSeats);
                    trip.setId(id);
                    trips.add(trip);
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return trips;
    }

    @Override
    public Optional<Trip> update(Trip trip) {
        // method to update a trip
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("update Trips set destination=?, dateDeparture=?, noOfAvailableSeats=? where id=?")) {
            preStmt.setString(1, trip.getDestination());
            LocalDateTime departureTime = trip.getDateDeparture();
            Timestamp timestamp = Timestamp.valueOf(departureTime);
            preStmt.setTimestamp(2, timestamp);
            preStmt.setLong(3, trip.getNoOfAvailableSeats());
            preStmt.setLong(4, trip.getId());
            int result = preStmt.executeUpdate();
            if (result > 0) {
                return Optional.of(trip);
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Trip> save(Trip trip) {
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("insert into Trips (destination, dateDeparture, noOfAvailableSeats) values (?,?,?)")) {
            preStmt.setString(1, trip.getDestination());
            LocalDateTime departureTime = trip.getDateDeparture();
            Timestamp timestamp = Timestamp.valueOf(departureTime);
            preStmt.setTimestamp(2, timestamp);
            preStmt.setLong(3, trip.getNoOfAvailableSeats());
            int result = preStmt.executeUpdate();
            if (result > 0) {
                try (Statement stmt = con.createStatement();
                     ResultSet rs = stmt.executeQuery("SELECT last_insert_rowid()")) {
                    if (rs.next()) {
                        trip.setId(rs.getLong(1));
                    }
                    else {
                        throw new SQLException("Creating trip failed, no ID obtained.");
                    }
                }
                return Optional.of(trip);
            }
        } catch (SQLException e) {
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Trip> delete(Long id) {
        // method to delete a trip by id
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        Optional<Trip> trip = findOne(id);
        if (trip.isPresent()) {
            try (PreparedStatement preStmt = con.prepareStatement("delete from Trips where id=?")) {
                preStmt.setLong(1, id);
                int result = preStmt.executeUpdate();
                if (result > 0) {
                    return trip;
                }
            } catch (SQLException e) {
                //logger.error(e);
                System.out.println("Error DB " + e);
            }
        }
        return Optional.empty();
    }
}
